File: services/service_request_service.py
from repositories.auth_repository import AuthRepository
from repositories.service_request_repository import ServiceRequestRepository
from utils.dto_mappers import to_service_request_dto
from utils.errors import ApiError
from utils.mongo_utils import parse_object_id
from utils.validators import require_fields, validate_string


class ServiceRequestService:
    def __init__(self):
        self.request_repo = ServiceRequestRepository()
        self.auth_repo = AuthRepository()

    def create_request(self, payload: dict, current_user: dict) -> dict:
        require_fields(payload, ["requestType", "requestDetail"])
        validate_string(payload["requestType"], "requestType")
        validate_string(payload["requestDetail"], "requestDetail")

        user_id = parse_object_id(current_user["userId"], "userId")
        user = self.auth_repo.get_user_by_id(user_id)
        if not user or user["role"] != "KHACHHANG":
            raise ApiError("Only customer can create service request", 403)

        request_doc = {
            "userId": user_id,
            "requestType": payload["requestType"],
            "requestDetail": payload["requestDetail"],
            "status": "MOI",
        }
        self.request_repo.create_request(request_doc)
        return to_service_request_dto(request_doc)

    def list_requests(self, current_user: dict) -> list[dict]:
        if current_user["role"] == "KHACHHANG":
            user_id = parse_object_id(current_user["userId"], "userId")
            docs = self.request_repo.list_requests_by_customer(user_id)
        else:
            docs = self.request_repo.list_requests()
        return [to_service_request_dto(doc) for doc in docs]

    def update_status(self, request_id: str, payload: dict) -> dict:
        require_fields(payload, ["status"])
        validate_string(payload["status"], "status")
        request_obj_id = parse_object_id(request_id, "requestId")
        existing = self.request_repo.get_request(request_obj_id)
        if not existing:
            raise ApiError("Service request not found", 404)

        self.request_repo.update_status(request_obj_id, payload["status"])
        updated = self.request_repo.get_request(request_obj_id)
        return to_service_request_dto(updated)

File: src/services/serviceRequestService.js
const { normalize } = require('../utils/legacyNormalizer');
const ServiceRequest = require('../models/ServiceRequest');
const User = require('../models/User');

class ServiceRequestService {
  async createServiceRequest(payload, currentUser) {
    // Normalize the payload (convert VI keys to EN keys)
    const normalized = normalize(payload, 'serviceRequest');

    // Extract fields
    const { requestType, requestDetail } = normalized;

    // Validate required fields
    if (!requestType || !requestDetail) {
      const error = new Error('Missing required fields');
      error.status = 400;
      throw error;
    }

    // Create service request document
    const serviceRequestData = {
      userId: currentUser.id,
      requestType,
      requestDetail,
      status: 'PENDING',
    };

    // Create service request
    const serviceRequest = await ServiceRequest.create(serviceRequestData);

    // Return DTO
    return this._toServiceRequestDTO(serviceRequest);
  }

  async getServiceRequests(currentUser) {
    // Build query based on user role
    let query = {};
    if (currentUser.role === 'KHACHHANG') {
      // Customers can only see their own requests
      query.userId = currentUser.id;
    }
    // Staff/Admin can see all requests (no filter needed)

    // Get service requests
    const serviceRequests = await ServiceRequest.find(query)
      .sort({ createdAt: -1 })
      .populate('userId', 'username fullName');

    // Convert to DTOs
    return serviceRequests.map(request => this._toServiceRequestDTO(request));
  }

  async updateServiceRequestStatus(requestId, status) {
    // Validate status
    const validStatuses = ['PENDING', 'IN_PROGRESS', 'RESOLVED', 'REJECTED'];
    if (!validStatuses.includes(status)) {
      const error = new Error('Invalid status');
      error.status = 400;
      throw error;
    }

    // Update service request
    const updatedRequest = await ServiceRequest.findByIdAndUpdate(
      requestId,
      { status },
      { new: true, runValidators: true }
    );

    if (!updatedRequest) {
      const error = new Error('Service request not found');
      error.status = 404;
      throw error;
    }

    // Return DTO
    return this._toServiceRequestDTO(updatedRequest);
  }

  // Helper method to convert service request to DTO (matches Python's to_service_request_dto)
  _toServiceRequestDTO(serviceRequest) {
    const sr = serviceRequest.toObject ? serviceRequest.toObject() : serviceRequest;

    return {
      // English
      id: sr._id.toString(),
      userId: sr.userId.toString(),
      requestType: sr.requestType,
      requestDetail: sr.requestDetail,
      status: sr.status,
      // Vietnamese
      MaYC: sr._id.toString(),
      MaKH: sr.userId.toString(),
      LoaiYeuCau: sr.requestType,
      ChiTietYeuCau: sr.requestDetail,
      TrangThai: sr.status,
    };
  }
}

module.exports = new ServiceRequestService();
